sources/packages/backend/apps/queue-consumers/src/services/cas-invoice-batch/cas-invoice-batch.service.ts

```diff
@@ -0,0 +1,250 @@
+import { Injectable } from "@nestjs/common";
+import { BC_TOTAL_GRANT_AWARD_CODE } from "@sims/services/constants";
+import {
+  CASDistributionAccount,
+  CASInvoice,
+  CASInvoiceBatch,
+  CASInvoiceBatchApprovalStatus,
+  CASInvoiceDetail,
+  CASInvoiceStatus,
+  DisbursementReceipt,
+  DisbursementValue,
+  DisbursementValueType,
+} from "@sims/sims-db";
+import { DataSource, EntityManager } from "typeorm";
+import { SequenceControlService, SystemUsersService } from "@sims/services";
+import { ProcessSummary } from "@sims/utilities/logger";
+
+/**
+ * Chunk size for inserting invoices. The invoices (and its details) will
+ * be grouped in chunks to be inserted in the database.
+ */
+const INVOICES_CHUNK_SIZE_INSERTS = 150;
+
+@Injectable()
+export class CASInvoiceBatchService {
+  constructor(
+    private readonly dataSource: DataSource,
+    private readonly sequenceControlService: SequenceControlService,
+    private readonly systemUsersService: SystemUsersService,
+  ) {}
+
+  /**
+   * Find e-Cert receipts that have no valid invoice associated,
+   * creating a new invoice batch to be sent to CAS if some
+   * pending receipts are found.
+   * @param processSummary process summary for logging.
+   * @returns new invoice batch and related data or null.
+   */
+  async createInvoiceBatch(
+    processSummary: ProcessSummary,
+  ): Promise<CASInvoiceBatch | null> {
+    return this.dataSource.transaction(async (entityManager) => {
+      let newBatchUniqueSequence: number;
+      // Consumes the next sequence number and prevent concurrent access.
+      await this.sequenceControlService.consumeNextSequenceWithExistingEntityManager(
+        "CAS_INVOICE_BATCH",
+        entityManager,
+        async (nextSequenceNumber: number) => {
+          newBatchUniqueSequence = nextSequenceNumber;
+        },
+      );
+      processSummary.info("Checking for pending receipts.");
+      const pendingReceipts = await this.getPendingReceiptsForInvoiceBatch(
+        entityManager,
+      );
```

Comment:
I believe that getting the pending receipts can be kept outside the transaction right? so that we can also save the rollback if there is not pending invoices to create.

Reply:
I am not sure if I followed this part "so that we can also save the rollback if there is not pending invoices to create.".
Either way, it is inside the transaction to share the connection but also to be part of the "blocked" execution.

Reply:
I still feel that fetching pending invoices can be outside the as it is the preliminary data which decides to either continue or abort the process.

Reply:
Got the reason now. thanks. resolving the comment.